File: android/lib/model/src/main/kotlin/com/warrenbrowse/vpn/lib/model/WarrenAnnouncement.kt
package com.warrenbrowse.vpn.lib.model

/**
 * One launch announcement, as Rust hands it over: already verified against the
 * pinned server key and already filtered for the envelope expiry, the
 * announcement's own TTL and this app's version.
 *
 * [headline] and [body] are the operator's own words and are rendered verbatim
 * as plain text. They are never translated and never parsed as markup: the
 * whole point of the signed channel is that what the operator wrote is what the
 * user reads.
 *
 * [voucherCampaignId] is the campaign this account may hold a code for, and its
 * presence IS the offer. [voucherCode] is that code once the wallet-signed
 * lookup has answered, `null` while it has not, and `null` for good when this
 * account is outside the cohort. It is a bearer token worth a month of service:
 * it belongs on the account's own screen and nowhere else, never in a log, an
 * error or a problem report.
 */
data class WarrenAnnouncement(
    val id: String,
    val headline: String,
    val body: String,
    val level: WarrenNoticeLevel,
    val cta: WarrenAnnouncementCta? = null,
    val voucherCampaignId: String? = null,
    val voucherCode: String? = null,
)

/**
 * The call to action an announcement carries, present only when the URL passed
 * the contract's own check in Rust. [label] is the operator's caption, [url] an
 * `https` destination opened in the system browser.
 */
data class WarrenAnnouncementCta(val label: String, val url: String)

File: desktop/packages/mullvad-vpn/src/renderer/lib/notifications/warren-announcement.ts
import { Url } from '../../../shared/constants';
import { WarrenAnnouncement, WarrenAnnouncementCta } from '../../../shared/daemon-rpc-types';
import {
  InAppAnnouncementCard,
  InAppNotification,
  InAppNotificationProvider,
} from '../../../shared/notifications';

interface WarrenAnnouncementNotificationContext {
  // Launch announcements the daemon verified against the pinned server key and
  // already filtered for expiry and this app's version. Each one is
  // display-ready, this account's voucher code included.
  announcements: WarrenAnnouncement[];
  // Announcement ids the user has put away, from the persisted GUI settings.
  dismissedIds: string[];
  dismiss: (id: string) => void;
}

// Operator launch announcement, rendered as a card on the connect screen.
//
// Ranked above the broadcast notice, which is the opposite of what their
// relative importance suggests, for a mechanical reason: a notice is not
// dismissible, so a long-lived one would bury the card for as long as it
// stands, and the card carries a code that stops being worth anything after
// the campaign. The card steps aside on its own the moment it is dismissed.
export class WarrenAnnouncementNotificationProvider implements InAppNotificationProvider {
  public constructor(private context: WarrenAnnouncementNotificationContext) {}

  public mayDisplay = () => this.current() !== undefined;

  public getInAppNotification(): InAppNotification {
    const announcement = this.current()!;
    const card: InAppAnnouncementCard = {
      id: announcement.id,
      body: announcement.body,
      voucherCode: announcement.voucherCode,
      cta: safeCta(announcement.cta),
      dismiss: () => this.context.dismiss(announcement.id),
    };
    return {
      indicator: indicatorFor(announcement.level),
      // The operator's own headline IS the title. Stacking a level word on top
      // of it would demote the words they wrote to a subtitle and say nothing
      // the indicator does not already say.
      title: announcement.headline,
      subtitle: announcement.body,
      action: { type: 'announcement-card', announcement: card },
    };
  }

  private current(): WarrenAnnouncement | undefined {
    return this.context.announcements.find(
      (announcement) => !this.context.dismissedIds.includes(announcement.id),
    );
  }
}

// The last gate before a control that opens a browser. The daemon already
// refused anything but https, and the envelope is signed, but this string
// still arrives from the network and the check costs nothing here.
export function announcementCtaUrl(cta: WarrenAnnouncementCta): Url | undefined {
  let parsed: URL;
  try {
    parsed = new URL(cta.url);
  } catch {
    return undefined;
  }
  return parsed.protocol === 'https:' ? (parsed.href as Url) : undefined;
}

function safeCta(cta: WarrenAnnouncementCta | null): InAppAnnouncementCard['cta'] {
  if (cta === null) {
    return null;
  }
  const url = announcementCtaUrl(cta);
  return url === undefined ? null : { label: cta.label, url };
}

function indicatorFor(level: WarrenAnnouncement['level']): InAppNotification['indicator'] {
  switch (level) {
    case 'error':
      return 'error';
    case 'warning':
      return 'warning';
    default:
      return 'success';
  }
}
